### ISSUE TITLE
Not compatible with Numpy 2.0

### ISSUE THREAD

**Issue by Timmmm** [networkx/networkx#7501]
Numpy 2.0 was just released and NetworkX is not compatible with it, but [does not restrict the version to `<2`](https://github.com/networkx/networkx/blob/197446d1bfa817588ebcf4269532cc33a033928e/pyproject.toml#L58). Importing `networkx` will give:

```
  File "yourcode.py", line 31, in <module>
    import networkx
  File "/venv/lib64/python3.11/site-packages/networkx/__init__.py", line 64, in <module>
    import networkx.readwrite
  File "/venv/lib64/python3.11/site-packages/networkx/readwrite/__init__.py", line 15, in <module>
    from networkx.readwrite.graphml import *
  File "/venv/lib64/python3.11/site-packages/networkx/readwrite/graphml.py", line 344, in <module>
    class GraphML:
  File "/venv/lib64/python3.11/site-packages/networkx/readwrite/graphml.py", line 377, in GraphML
    (np.float_, "float"),
     ^^^^^^^^^
  File "/venv/lib64/python3.11/site-packages/numpy/__init__.py", line 397, in __getattr__
    raise AttributeError(
AttributeError: `np.float_` was removed in the NumPy 2.0 release. Use `np.float64` instead.. Did you mean: 'float16'?
```

**Comment by Schefflera-Arboricola** [comment]
I am not getting this error while importing networkx (v3.3) with numpy 2.0. I wonder which version of networkx you are using.

These numpy types were added in networkx 2.2 (ref PR https://github.com/networkx/networkx/pull/3041) and then `np.float_` was removed in networkx 3.2 (ref PR https://github.com/networkx/networkx/pull/6900)

And currently there is no `(np.float_, "float")` in the `graphml` code:
https://github.com/networkx/networkx/blob/197446d1bfa817588ebcf4269532cc33a033928e/networkx/readwrite/graphml.py#L414-L418

Thank you :)

**Comment by Timmmm** [comment]
Ah looks like we are actually using `networkx==2.5.1`, which is in that range. I did check that Pyproject.toml allows any future version:

```
    'numpy>=1.23',
```

but I guess the current code is lucky not to use anything that broke in 2.0 (at least at import time).

The proper fix is to use semver instead (`numpy^1.23`) but apparently Numpy doesn't even use semver so I guess there is nothing you can do and we'll have to put up with occasional breakage. :disappointed:
